types/evidence.go

```diff
@@ -0,0 +1,114 @@
+package types
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/tendermint/go-crypto"
+	"github.com/tendermint/tmlibs/merkle"
+)
+
+// ErrEvidenceInvalid wraps a piece of evidence and the error denoting how or why it is invalid.
+type ErrEvidenceInvalid struct {
+	Evidence   Evidence
+	ErrorValue error
+}
+
+func NewEvidenceInvalidErr(ev Evidence, err error) *ErrEvidenceInvalid {
+	return &ErrEvidenceInvalid{ev, err}
+}
+
+// Error returns a string representation of the error.
+func (err *ErrEvidenceInvalid) Error() string {
+	return fmt.Sprintf("Invalid evidence: %v. Evidence: %v", err.ErrorValue, err.Evidence)
+}
+
+//-------------------------------------------
+
+// Evidence represents any provable malicious activity by a validator
+type Evidence interface {
+	Address() []byte
+	Hash() []byte
+	Verify(chainID string) error
+
+	String() string
+}
+
+//-------------------------------------------
+
+type Evidences []Evidence
+
+func (evs Evidences) Hash() []byte {
+	// Recursive impl.
+	// Copied from tmlibs/merkle to avoid allocations
+	switch len(evs) {
+	case 0:
+		return nil
+	case 1:
+		return evs[0].Hash()
+	default:
+		left := Evidences(evs[:(len(evs)+1)/2]).Hash()
+		right := Evidences(evs[(len(evs)+1)/2:]).Hash()
+		return merkle.SimpleHashFromTwoHashes(left, right)
+	}
+}
+
+//-------------------------------------------
+
+// DuplicateVoteEvidence contains evidence a validator signed two conflicting votes.
+type DuplicateVoteEvidence struct {
+	PubKey crypto.PubKey
+	VoteA  *Vote
+	VoteB  *Vote
+}
+
+// String returns a string representation of the evidence.
+func (dve *DuplicateVoteEvidence) String() string {
+	return fmt.Sprintf("VoteA: %v; VoteB: %v", dve.VoteA, dve.VoteB)
+
+}
+
+// Address returns the address of the validator.
+func (dve *DuplicateVoteEvidence) Address() []byte {
+	return dve.PubKey.Address()
+}
+
+// Hash returns the hash of the evidence.
+func (dve *DuplicateVoteEvidence) Hash() []byte {
+	return merkle.SimpleHashFromBinary(dve)
+}
+
+// Verify returns an error if the two votes aren't conflicting.
+// To be conflicting, they must be from the same validator, for the same H/R/S, but for different blocks.
+func (dve *DuplicateVoteEvidence) Verify(chainID string) error {
+	// H/R/S must be the same
+	if dve.VoteA.Height != dve.VoteB.Height ||
+		dve.VoteA.Round != dve.VoteB.Round ||
+		dve.VoteA.Type != dve.VoteB.Type {
+		return fmt.Errorf("DuplicateVoteEvidence Error: H/R/S does not match. Got %v and %v", dve.VoteA, dve.VoteB)
+	}
+
+	// Address must be the same
+	if !bytes.Equal(dve.VoteA.ValidatorAddress, dve.VoteB.ValidatorAddress) {
+		return fmt.Errorf("DuplicateVoteEvidence Error: Validator addresses do not match. Got %X and %X", dve.VoteA.ValidatorAddress, dve.VoteB.ValidatorAddress)
+	}
+	// XXX: Should we enforce index is the same ?
```

Comment:
should we? checking the address should be enough. although, validator set is sorted, so it does not matter